### PR TITLE
lib-user: Refactor GroupStats for user, group, and join states

### PR DESCRIPTION
**Pull request by mcbouslog**
## Package
- app-root (small change)
- lib-user (mostly)

## Describe your changes
- pass join_token url param to GroupStats component
- move stats requests and related projects request from GroupStatsContainer to GroupStats
- refactor user_group create and update for consistency
- refactor user_group hook per join status
- refactor GroupStatsContainer and create group membership per joinToken
- refactor GroupStatsContainer per user, group, and join status

## How to Review
_Helpful explanations that will make your reviewer happy:_
- What Zooniverse project should my reviewer use to review UX? n/a
- What user actions should my reviewer step through to review this PR?
  - production (from app-root, `yarn build`, `yarn start`):
    - logged out
      - [x] private group (https://local.zooniverse.org:3000/groups/2597524) = can't find group please login message
      - [x] with join_token in url (https://local.zooniverse.org:3000/groups/2597524?join_token=1234567) = please login message
    - logged in
      - [x] private group not a member of (https://local.zooniverse.org:3000/groups/2597524) = can't find group message
      - [x] private group a member of = confirm an existing group you're a member of works as expected
      - [x] try to join a group with an incorrect join_token (https://local.zooniverse.org:3000/groups/2597524&join_token=1234567) = join error message
      - [x] join a group with a correct join_token (https://local.zooniverse.org:3000/groups/2597524?join_token= 7939f9717b253c03) 
  - staging (from lib-user, `yarn dev`): 
    - logged out
      - [x] private group (https://local.zooniverse.org:8080/?groups=1326962) = can't find group please login message
      - [x] public group loads (https://local.zooniverse.org:8080/?groups=1326971)
      - [x] with join_token in url (https://local.zooniverse.org:8080/?groups=1326971&join_token=1234567) = please login message
    - logged in
      - [x] private group not a member of (https://local.zooniverse.org:8080/?groups=1326962) =  can't find group please login message
      - [x] private group a member of = confirm an existing group you're a member of works as expected
      - [x] try to join a group with an incorrect join_token (https://local.zooniverse.org:8080/?groups=1326971&join_token=1234567) = join error message
      - [x] join a group with a correct join_token (https://local.zooniverse.org:8080/?groups=1326962&join_token=59269f7cbd67bdc1)
- Which storybook stories should be reviewed? n/a
- Are there plans for follow up PR’s to further fix this bug or develop this feature? yes, see linked GitHub project, notably:
  - show/don't show specific components, like certain header items (i.e. Manage Group) and certain components (i.e. Top Contributors) per group and user roles

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [ ] Tests are passing locally and on Github
- [ ] Documentation is up to date and changelog has been updated if appropriate
- [ ] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [ ] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## New Feature
- [ ] The PR creator has listed user actions to use when testing the new feature
- [ ] Unit tests are included for the new feature

## Refactoring
- [ ] The PR creator has described the reason for refactoring
- [ ] The refactored component(s) continue to work as expected

### CODE REVIEW

File: packages/lib-user/src/components/GroupStats/EditGroup.js [hunk 1]
```diff
@@ -4,42 +4,70 @@ import { useEffect, useState } from 'react'
 
 function EditGroup({ group, handleGroupUpdate }) {
```
Comment (mcbouslog) [inline packages/lib-user/src/components/GroupStats/EditGroup.js:5]: Noting the EditGroup component will be replaced with an iteration on the MyGroups/components/GroupForm, nonetheless it helps to have the following for now.

File: packages/lib-user/src/components/GroupStats/GroupStatsContainer.js [hunk 2]
```diff
@@ -1,32 +1,39 @@
...
 
 import {
+  createPanoptesMembership,
   deletePanoptesUserGroup,
   getBearerToken,
-  getDateInterval,
   updatePanoptesUserGroup
 } from '@utils'
 
+import { getStatus } from './helpers/getStatus'
```
Comment (mcbouslog) [inline packages/lib-user/src/components/GroupStats/GroupStatsContainer.js:20]: I don't love the name of this helper function, open to suggestion.

Comment (goplayoutside3) [reply packages/lib-user/src/components/GroupStats/GroupStatsContainer.js:20]: Would `getJoinStatus` or `getUserGroupStatus` make sense?

File: packages/lib-user/src/utils/createPanoptesMembership.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    { memberships: {
+      join_token: joinToken,
+      links: {
+        user: userId,
+        user_group: groupId
+      }
+    }},
+    { authorization }
+  )
+  return response
```
Comment (goplayoutside3) [inline packages/lib-user/src/utils/createPanoptesMembership.js:23]: Just to double check - do the `utils` need some basic error handling? For instance, `getBearerToken()` util has a try/catch, but the other functions do not, and they all return `response` from `panoptes` regardless. There may be some error handling already in `panoptes` from `@zooniverse/panoptes-js`, but good to double check at this PR!

**Review by goplayoutside3 — COMMENTED** [review]
I'm able to follow the UI testing steps, however these two I could not test because 2597524 is a private group.

>  - try to join a group with an incorrect join_token (https://local.zooniverse.org:3000/groups/2597524&join_token=1234567) = join error message
> - join a group with a correct join_token (https://local.zooniverse.org:3000/groups/2597524&join_token=7939f9717b253c03)

I was able to join your public group linked in the lib-user testing section. However, when I tried to join the same group again (https://local.zooniverse.org:8080/?groups=1326962&join_token=59269f7cbd67bdc1), I got the message "New group joined!", and then the UI auto refreshes to show "Group Not Found". Is that expected? I thought if I try to join a group I'm already a member of, the page should just show me the group, and not try to resolve the token again.

File: packages/lib-user/src/components/GroupStats/GroupStatsContainer.js (hunk 2 above)
Comment (mcbouslog) [reply packages/lib-user/src/components/GroupStats/GroupStatsContainer.js:20]: I like `getUserGroupStatus` better than just `getStatus` 👍

**Comment by mcbouslog** [comment]
> I'm able to follow the UI testing steps, however these two I could not test because 2597524 is a private group.
> 
> > * try to join a group with an incorrect join_token (https://local.zooniverse.org:3000/groups/2597524&join_token=1234567) = join error message
> > * join a group with a correct join_token (https://local.zooniverse.org:3000/groups/2597524&join_token=7939f9717b253c03)
> 

🤦‍♂️  I did not have the `join_token` query param properly configured for `app-root`. The subpaths are included in `params`, but URL query params, like `join_token` are included in `searchParams`. I've updated the code accordingly.

**Comment by mcbouslog** [comment]
> I was able to join your public group linked in the lib-user testing section. However, when I tried to join the same group again (https://local.zooniverse.org:8080/?groups=1326962&join_token=59269f7cbd67bdc1), I got the message "New group joined!", and then the UI auto refreshes to show "Group Not Found". Is that expected? I thought if I try to join a group I'm already a member of, the page should just show me the group, and not try to resolve the token again.

The noted UX is not intended, I've added a request for a group membership and conditions so if try to join and already a member it doesn't try to create another membership, as well as code so that after joining the group should load as expected.

File: packages/lib-user/src/utils/createPanoptesMembership.js (hunk 3 above)
Comment (mcbouslog) [reply packages/lib-user/src/utils/createPanoptesMembership.js]: 👍 Added console.error's to related resource utils. Still throwing the error for consuming function/component to do as necessary with, as opposed to getBearerToken which I think expectedly captures the error, I think with these resource utils we want the error to be thrown, so slight difference with throw/capture error intended (though let me know if mistaken in understanding).

**Comment by goplayoutside3** [comment]
I can confirm that when trying to join a group I'm already in, the UI just shows me the group 👍 

Could you double check this url for your group through app-root? 
> - join a group with a correct join_token (https://local.zooniverse.org:3000/groups/2597524&join_token=7939f9717b253c03)

Whenever I click that url, the group is always "not found". Should I be able to join it when running app-root locally and signed into my staging account?

**Comment by mcbouslog** [comment]
> * https://local.zooniverse.org:3000/groups/2597524&join_token=7939f9717b253c03

Apologies I had the url incorrect, it should be https://local.zooniverse.org:3000/groups/2597524?join_token=7939f9717b253c03 and it's for a user_group on production. Here's a different production user_group for testing as well - https://local.zooniverse.org:3000/groups/2619372?join_token=24a155bb4c59c5fa .

Alternatively here's a staging user_group with join_token link for app-root pointed to staging (i.e. app-root `yarn dev`) - https://local.zooniverse.org:3000/groups/1326974?join_token=a31a20a6fea05d55 .

**Comment by goplayoutside3** [comment]
> Alternatively here's a staging user_group with join_token link for app-root pointed to staging (i.e. app-root yarn dev) - https://local.zooniverse.org:3000/groups/1326974?join_token=a31a20a6fea05d55 .

I was able to join this one! I think I was previously running `yarn dev` every time, so the production group token wasn't working.

When I visit a group url with a join token on app-root (either in dev mode or production mode) I do see the "You've joined a new group" alert 👍 However, the page itself shows "Group not found". The page will eventually reload with the actual group data via SWR triggers, but should something be done to either 1) delay page render until a token is successfully used to join the group or 2) force a re-render upon user successfully joining the group? 🤔 There's a bit of a mismatch if a user clicks the join link and sees "Group not found" even though they've successfully joined.

Let me know if you need a screenshot of the above behavior.

**Comment by mcbouslog** [comment]
Yes, I saw the same UX. I think I've refactored the related memberships and user_groups requests accordingly, so now when a user successfully joins it loads the group as expected. Here are two new groups to test with, noting the issue you note is (I think) only noticeable on a private group, I've included a new public group as well:
staging user groups for testings:
- private group: https://local.zooniverse.org:3000/groups/1326984?join_token=62403a6204292768
- public group: https://local.zooniverse.org:3000/groups/1326985?join_token=b0b664642bb9a23b

**Review by goplayoutside3 — APPROVED** [review]
Nice! Now when I join a group, I see the actual group data ✅

**Pull request merged**